developing/vcs/README.md

```diff
@@ -2,7 +2,321 @@
+multiple commits, some of which might be broken.
+
+Ways to mitigate cons:
+
+* There is a way to squash commits locally without `rebase`, but it involves
+[several steps](https://gist.github.com/n00neimp0rtant/9515611), the last of
+which requires a force push, which defeats the purpose of using `merge`, and
+does not allow reordering or choosing more than one commit.
+
+* You can use `merge` locally, and then squash and reorder commits on GitHub
+right before merging, but that forces you to delete local branches with `-D`,
+which we don't think anyone should have to do.
+
+#### Rebase
+
+Pros:
+
+* Cleaner, more meaningful, easier to understand commit history
+* Easier to keep only commits that pass tests, which makes debugging
+easier with tools like `git bisect`.
+* Safer and easier to handle conflicts when commits are squashed
+* Easier to revert bad commits
+* Easier to change commit messages in a series of commits. We value great commit
+messages, so it should be easy to improve any message in a series of commits.
+* Allows reordering commits for more meaningful grouping. For example, you might
+have a PR with several commits that each update a dependency. If the tests fail
+due to a dependency in the middle, you'll want to fix that in a separate commit,
+then rebase interactively to move the fix under the dependency commit and
+combine it with `fixup`, then edit the commit message to explain what had to
+change after updating the dependency.
+
+Cons:
+
+* Riskier since it requires a force push
+* Rewrites history
+
+Ways to mitigate cons:
+
+* Protect the master branch from force pushes
+* Agree on a process where once a PR is under active review, no force pushes
+until the PR is approved and ready to be merged.
+* Original history can be viewed with `git reflog`
+
+#### Examples of popular repos that enforce rebase and squash
+
+* [Rails](https://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#squashing-commits)
+* [Django](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/working-with-git/#rebasing-branches)
+* [Golang](https://golang.org/doc/contribute.html#make_branch)
+* [Homebrew](https://docs.brew.sh/How-To-Open-a-Homebrew-Pull-Request#following-up)
+* [Hugo](https://github.com/gohugoio/hugo/blob/master/CONTRIBUTING.md#code-contribution-guidelines)
+* [React](https://github.com/facebook/react)
+* [Bootstrap](https://github.com/twbs/bootstrap/blob/master/.github/CONTRIBUTING.md#pull-requests)
+* [Electron](https://electronjs.org/docs/development/pull-requests#step-6-rebase)
+* [Node](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#step-5-rebase)
+
+#### Examples of popular repos that enforce merge
+
+I'm trying to avoid confirmation bias, but I haven't found any yet. If you know
+of any, please add them. I found some that don't enforce any specific workflow,
+and most of them say they will squash your commits via the GitHub interface.
```

Comment:
Linux uses merge: [Linus on Merge vs Rebase](https://web.archive.org/web/20181107065817/http://thread.gmane.org/gmane.comp.video.dri.devel/34739/focus=34744), [kernel.org contributing docs](https://www.kernel.org/doc/html/latest/maintainer/rebasing-and-merging.html)

Reply:
Hmm. The second link doesn't say that they specifically prefer merge. They talk about the benefits of both. On GitHub, they point people to this document, which specifically ask people to either split their work into separate logical patches, or to squash commits using interactive rebase: https://www.kernel.org/doc/html/latest/process/submitting-patches.html

Reply:
This is that thing again where we are fighting the battle on the wrong field. The conflict isn't really about merging and rebasing. The contention, I think, is force pushing in a way that destroys published history, including (potentially) others' work. I don't think anybody objects to rebasing work that hasn't been published anywhere.

Reply:
I definitely don't object to rebasing unpublished work. I use rebase from time to time, but I do think that it's a pretty sharp edged tool to be using as part of our recommended default flow, recommended to junior engineers who have a lot more important thinks to be worrying about than the intricacies of git.

Reply:
Remember the context is finding projects that use a merge workflow and eschew rebasing. I think Linux is the ultimate case of that, and should be mentioned if we will have a list.

From the kernel.org docs: 

> History that has been exposed to the world beyond your private system should usually not be changed...If work is in need of rebasing, that is usually a sign that it is not yet ready to be committed to a public repository.

On my reading, that uses a much more contrite tone but basically paraphrases and condenses a big part of Linus' email. An exception is mentioned but it's strictly confined to when git is being used to distribute work in progress, _not_ for others to base their own work on (since destroying published history is bad).

Here's the clear statement of intent, declaring allegiance to merge:

> Many projects require that branches in pull requests be based on the current trunk so that no merge commits appear in the history. The kernel is not such a project; any rebasing of branches to avoid merges will, most likely, lead to trouble.

The submitting patches doc where `rebase -i` is encouraged seems to be about creating pull requests that are likely to be accepted, but not rewriting published history.

Reply:
I'm realizing now that we might be talking about different things. There are two different concepts: one is rebasing in order to tidy up a PR *before* it is merged, and the other is how the PR ends up being merged into `master`. They are not the same thing. The way I read the second quote above deals with the second concept: how a branch/PR gets incorporated into the trunk (such as the `master` branch). GitHub provides 3 different ways to do that: merge commit, squash, and rebase. 

When the quote above refers to no merge commits appearing in the history, I understand that to be the equivalent of using the rebase button on GitHub, which does not add a merge commit (the squash button as well I believe does not add a merge commit).

The workflow I am proposing is that we always use the merge option on GitHub to merge a PR (which is what I had all along in the `GitHub repo settings` section), but if the PR has commits with poor messages or that are just fixing linting errors, or that cause tests to fail, those should be cleaned up first.

When I started that list of popular repos, I did not frame it correctly. I was only looking at their policy for preparing PRs, not how the PR is merged. I think that most people agree that it's a good idea to clean up PRs before they get merged, including Linux.

I will create a more specific poll to see what people think about cleaning up PRs and which buttons they prefer on GitHub.

Reply:
> I'm realizing now that we might be talking about different things. There are two different concepts: one is rebasing in order to tidy up a PR before it is merged, and the other is how the PR ends up being merged into master.

Yes. As far as I'm concerned this is about `push --force[-with-lease]`. 

>  I think that most people agree that it's a good idea to clean up PRs before they get merged, including Linux.

They do. In fact, they believe in it so much, they require that the PR should be cleaned up before it is even sent! But once it's sent, the history is published and should not be changed.

> History that has been exposed to the world beyond your private system should usually not be changed...If work is in need of rebasing, that is usually a sign that it is not yet ready to be committed to a public repository.

I think that means Linux & kernel.org would go in this part of the list.

Reply:
I try to follow Charlie Munger's advice about [fully understanding the other side's argument](https://fs.blog/2013/04/the-work-required-to-have-an-opinion/), but I don't think I do yet. Here is how I understand what you said. Please let me know where I misunderstood.

In the Linux example:
- You are equating submitting a patch with it being published, regardless of whether or not it is accepted. In other words, if someone submits a patch and the reviewer asks for changes, and if the submitter then creates several new commits locally to address the feedback, they can no longer rebase those commits to tidy them up before resubmitting.
- You are equating submitting a patch to Linux to submitting a PR to a repo in GitHub.

Is that what you are saying?

If so, that's not how I understand it, and I'd like to know where I might be wrong. The way I read the Linux contribution instructions [here](https://github.com/torvalds/linux/pull/415#issuecomment-301877514) is that a patch is sent via `git send-email`, and if it requires changes, another revision of the patch is sent using the version and annotate flags, such as `git send-email -v2 --annotate`, and the contributor describes what changed between versions 1 and 2.

In the GitHub PR above, the Linux maintainer tells the contributor that it's [fine to amend their commit](https://github.com/torvalds/linux/pull/415#issuecomment-302276507). The 2 patches are documented here:

[PATCH 1](https://marc.info/?l=linux-mm&m=149502833306198&w=2)
[PATCH v2](https://marc.info/?l=linux-mm&m=149504876613136&w=2)

The second patch doesn't include the original commit. It is an amended version of the first patch.

To me, that is the equivalent of using force push once a PR is approved, since that is the only way of updating the PR in order to do a final cleanup after addressing changes requested during PR review.

While a Linux patch is still being reviewed via email, I don't consider it to be published publicly. Only once it is merged into the "subsystem tree" would I consider it "exposed to the world".

In the world of GitHub, I do consider a PR to be published publicly, and I fully understand the risk of force pushing to a GitHub PR that other folks might be actively working on. What isn't yet clear to me is why folks think it's also risky to force push even if it's only done once at the end of a PR's life, when no one else will be touching that branch anymore.

To me, there isn't much difference between 1) pairing with someone locally and pushing the final PR and 2) pushing to GitHub first to have someone review the code remotely, making changes, and then force pushing to update the final commit message to document any important points that came up during the review.

The end goal should be to have a detailed commit message that also captures any relevant comments that might have been made during review. As I mentioned in Slack, I don't think it's a good idea to rely on GitHub as our canonical source of documentation. If there is relevant information in the comments of a PR, it should be summarized and captured in the final commit message. One shouldn't have to visit GitHub and try to read through all the comments, some or most of which might not be valuable, in order to get the big picture. I realize that the PR's main body can be updated on GitHub to summarize everything, but I don't think one should have to visit a website to understand a commit. If I want to look up a commit locally, I want to be able to stay in my Terminal or editor and not have to switch to a browser.

If we all agree that it's a good idea to document the PR in the commit message as opposed to on GitHub, and that it's OK to force push when it's necessary to update the final commit message, then I think we have a middle ground:

- those who prefer rebase can use it consistently, both to incorporate changes from the main branch, and to edit commits
- those who prefer merge can use it consistently, both to incorporate changes from the main branch, and to edit the final commit message using `git merge --squash`

In both cases, a force push is required at the end, and in both cases, we will use the "Merge" button on GitHub to incorporate the PR into the main branch. How does that sound?

Reply:
@monfresh I interpret the PATCH and PATCHv2 as... not like rebase and force push; conceptually, if you add a commit (PATCHv2), you’re not rebasing, you’re pushing a new commit into your branch.

I see what appears to be a consistent misunderstanding of the concerns around rebase: the problem isn’t “is this safe when everything goes right”, it’s “what happens when this doesn’t go right”. This is the fundamental concern with git workflows that require force push anywhere in the flow; at some point, someone’s going to typo a branch and have a bad day.

I’m actually ok with us accepting that risk if we think the trade off of a clean history (which, TBH, I think is over-valued, however, that’s a belief, not a fact). I am *not* ok with us having two ways to do the basic, every day, 100% needed operation. That seems ridiculous.

So: here’s what I want: do the risk analysis, with a focus on correctness and safety, in a way that (in theory) should convince this merry band of *engineers*  that it’s better to have one way to do this that we train and socialize, even if it’s not the way we currently believe to be best.